### PR TITLE
Add support for env hashing

### LINKED ISSUE #401: Caching environment variables on Vercel
### What version of Turborepo are you using?

1.0.19

### Describe the Bug

On Vercel I have a public environment variable set called `NEXT_PUBLIC_SEGMENT_WRITE_KEY`. It's set to one value for production, and another for preview and development.

If I deploy a preview branch with the build command set to `cd ../.. && yarn turbo run build --scope=website --includeDependencies --no-deps` and preview the site, the correct env var for preview/dev used.

But then if I then do a production deployment, the env var for preview/dev is still used. The env var is used within a function, within a component like this:

```js
  const renderSnippet = () => {
    const opts = {
      apiKey: process.env.NEXT_PUBLIC_SEGMENT_WRITE_KEY,
      load: true,
      page: true,
    };

    if (!opts.apiKey) {
      return null;
    }

    return snippet.min(opts);
  };
```

Clearing the build command so it's just the Next.js default restores the normal behaviour.

### Expected Behavior

Correct production env var is picked up and used.

### To Reproduce

1. Create a public environment variable with different values for production and preview/development.
2. Render `process.env.NEXTJS_PUBLIC_<var name>` to page.

### PR DESCRIPTION
**Pull request by jaredpalmer**
Fix #401 

This adds a new `env` key to the root config and also the `pipeline`. 

- When specified at the root, `turbo` will hash the key-values of the listed env vars and include the resulting list in the global hash value which is included in all other package task hashes. This is useful for having dev/prod/local caches.
- When specified in the pipeline, `turbo` will hash  the key-values of the listed env vars and include the resulting sorted list in the applicable task hash. This is ideal for Next.js/Remix build tasks that inline environment variables. 

New docs: 
- https://turbo-site-git-add-hashable-env-vars.vercel.sh/docs/features/caching#alter-caching-based-on-environment-variables
- https://turbo-site-git-add-hashable-env-vars.vercel.sh/docs/features/caching#hashing
- https://turbo-site-git-add-hashable-env-vars.vercel.sh/docs/reference/configuration#env
- https://turbo-site-git-add-hashable-env-vars.vercel.sh/docs/reference/configuration#env-1

**Example**
```json
{
  "turbo": {
    "baseBranch": "origin/master",
    "env": [ // values will impact the hashes of all tasks
      "GITHUB_TOKEN"
    ],
    "pipeline": {
      "build": {
        "dependsOn": {
          "^build"
        },
        "outputs": ["dist/**"]
        "env": [ // values will impact hashes of all "build" tasks
          "SOME_ENV_VAR"
        ]
      },
      "web#build": { // override settings for the "build" task for the "web" package
        "dependsOn": [
          "^build"
        ],
        "outputs": [".next/**"],
        "env": [ // values will impact the hash of "build" task for only "web" package
           "STRIPE_SECRET_KEY",
           "NEXT_PUBLIC_STRIPE_PUBLIC_KEY",
           "NEXT_PUBLIC_ANALYTICS_ID",
        ]
      }
    },
  }
}
```

### CODE REVIEW

File: docs/pages/docs/features/caching.mdx [hunk 1]
```diff
@@ -75,6 +75,48 @@ You can also always disable caching on a specific task by setting `cache` to `fa
 }
 ```
 
+## Alter Caching Based on Environment Variables
+
+You can control caching based on the value of environment variables by setting `env` key where applicable.
+
+**At the root** will tell `turbo` to include the hashed values of each environment variable it all other hashes.
+**In the `pipeline`** will tell `turbo` to inclued the hashed values of each environment variable on a per-task or per-package-task basis.
```
Comment (gaspar09) [inline docs/pages/docs/features/caching.mdx:83]: `inclued` -> `include`

File: docs/pages/docs/reference/configuration.mdx [hunk 2]
```diff
@@ -19,6 +19,34 @@ Defaults to `origin/master`. The base branch or your git repository. Git is used
 A list of globs for implicit global hash dependencies. The contents of these files will be included in the global hashing algorithm.
 This is useful for busting the cache based on `.env` files (not in Git) or any root level file that impacts package tasks (but are not represented in the traditional dependency graph (e.g. a root `tsconfig.json`, `jest.config.js`, `.eslintrc`, etc.)).
 
+## `env`
+
+`type: string`
+
+A list of the names of the environment variables that should be included in the global hashing function (which impacts _all_ tasks).
```
Comment (gaspar09) [inline docs/pages/docs/reference/configuration.mdx:26]: Maybe simplify to:

`A list of the names of the environment variables...` -> `A list of environment variables...`

File: cli/internal/run/run.go [hunk 3]
```diff
@@ -419,16 +419,27 @@ func (c *RunCommand) Run(args []string) int {
 					outputs = append(outputs, pipeline.Outputs...)
 				}
 				targetLogger.Debug("task output globs", "outputs", outputs)
+				// Hash the task-specific environment variables
+				var hashabledEnvVars = make([]string, 0, len(pipeline.HashedEnv))
```
Comment (gaspar09) [inline cli/internal/run/run.go:423]: Should this be `hashableEnvVars`

File: docs/pages/docs/features/caching.mdx [hunk 4]
```diff
@@ -75,6 +75,48 @@ You can also always disable caching on a specific task by setting `cache` to `fa
 }
 ```
 
+## Alter Caching Based on Environment Variables
+
+You can control caching based on the value of environment variables by setting `env` key where applicable.
+
+**At the root** will tell `turbo` to include the hashed values of each environment variable it all other hashes.
```
Comment (weyert) [inline docs/pages/docs/features/caching.mdx:82]: As non native speaker I find this confusing wording, I find the bit 'the hashed values of each environment variable it all other hashes' confusing.

File: docs/pages/docs/features/caching.mdx [hunk 5]
```diff
@@ -75,6 +75,48 @@ You can also always disable caching on a specific task by setting `cache` to `fa
...
+    "env": [ // values will impact the hashes of all tasks
+      "GITHUB_TOKEN"
+    ],
+    "pipeline": {
+      "build": {
+        "dependsOn": {
+          "^build"
+        },
+        "outputs": ["dist/**"]
+        "env": [ // values will impact hashes of all "build" tasks
```
Comment (weyert) [inline docs/pages/docs/features/caching.mdx:98]: Does this combine the values of `turbo.env` with the ones defined in the task or are they getting override? Would this end up with `["GITHUB_TOKEN", "SOME_ENV_VAR"]` or `["SOME_ENV_VAR"]`?

File: docs/pages/docs/features/caching.mdx (hunk 4 above)
Comment (weyert) [reply docs/pages/docs/features/caching.mdx:82]: I will here repeat the example that's given on a different page, this one:

>  - The value of the GITHUB_TOKEN environment variable will alter the hash of all tasks.
>  - The values of GITHUB_TOKEN and SOMETHING_ELSE environment variables will alter the hash of all build tasks
>  - The values of GITHUB_TOKEN, STRIPE_SECRET_KEY, and NEXT_PUBLIC_STRIPE_PUBLIC_KEY environment variables will alter the hash of only the build task for a package/app named web
>

**Comment by jaredpalmer** [comment]
Closing in favor of #495

**Pull request closed without merging**